addon/utils/handle-key-event.js

```diff
@@ -0,0 +1,52 @@
+import Ember from 'ember';
+import KEY_MAP from 'ember-keyboard/fixtures/key-map';
+
+const { hasListeners } = Ember;
+
+// Transforms jquery events' `keyup` and `keydown` into Ember conventional `keyUp` and `keyDown`.
+const normalizeEventType = function normalizeEventType(key, event) {
+  return `key${Ember.String.capitalize(event.type.replace('key', ''))}:${key}`;
+};
+
+// Since app devs might define their modifier keys in any order (eg. `keyUp:shift+ctrl+a` or
+// `keyUp:ctrl+shift+a`), we must check for each variant until we find the correct one.
+const gatherEventNameVariants = function gatherEventNameVariants(event, keys, eventName) {
+  if (keys.length === 0) {
+    return [eventName];
+  }
+
+  return keys.reduce((variants, key) => {
+    const modifiedEventName = !eventName ? normalizeEventType(key, event) : `${eventName}+${key}`;
+    const remainingKeys = keys.filter((keyName) => keyName !== key);
+
+    return variants.pushObjects(gatherEventNameVariants(event, remainingKeys, modifiedEventName));
+  }, Ember.A());
+};
+
+// Check if any modifier keys are being held down.
+const gatherKeys = function gatherKeys(event) {
+  const key = event.key || KEY_MAP[event.keyCode];
+
+  return ['ctrl', 'meta', 'alt', 'shift'].reduce((keys, keyName) => {
+    if (event[`${keyName}Key`]) {
+      keys.pushObject(keyName);
+    }
+
+    return keys;
+  }, Ember.A([key]));
+};
+
+export default function handleKeyEvent(event, responderStack) {
+  let triggeredVariant;
+  const keys = gatherKeys(event);
+  const variants = gatherEventNameVariants(event, keys);
+
+  // Finds the first responder with a registered listener for one of the variants
+  const responder = responderStack.find((responder) => {
+    return triggeredVariant = variants.find((variant) => hasListeners(responder, variant));
```

Comment:
Yeah, I'm not happy with this either, but can't think of a better approach. I've got another branch we're I'm exploring using web workers to take over this util, but the main challenge is actually doing exactly what you said: getting the names of the responders, or rather, the listeners. We can dive deep into the internals of `__ember_meta__._listeners` to peek at them, but as of 2.0, even what we find there is pretty gnarly. Does not look like a stable api to tie into. If you've got a moment tomorrow, it'd be great to sit with you and go over that stuff. Maybe we can find a way to use it.